File: backend/src/scheduler/core/preference.ts
import {
  PREFERENCE_MATRIX_LENGTH,
  PREFERENCE_SLOTS_PER_DAY,
} from "@zenflow/shared";
import { clamp, utcToMinutes } from "../../common/utils";
import { PREFERENCE_LEARNING_RATE } from "../constants";
import { dragDistanceReward } from "./reward";
import { isoWeekday, localDateStr } from "./slot";

/**
 * Preference-matrix helpers — pure, no I/O, no clock, no randomness
 * (CLAUDE.md invariant 2). The matrix is a flat **168** signed floats:
 * 7 ISO weekdays × 24 one-hour buckets, row-major by weekday
 * (`@zenflow/shared`'s `PREFERENCE_MATRIX_LENGTH` / `PREFERENCE_SLOTS_PER_DAY`).
 * Slot scoring built on these lives in `slot-score.ts`.
 */

/**
 * Flat-array index into the 7×24 signed preference matrix, 7 ISO weekdays
 * (Mon = 1 … Sun = 7) × 24 one-hour buckets, row-major by weekday.
 */
export function matrixIndex(isoWeekdayNum: number, hour: number): number {
  return (isoWeekdayNum - 1) * PREFERENCE_SLOTS_PER_DAY + hour;
}

/**
 * Cold-start default population, used whenever a user's stored
 * `preferenceMatrix` is empty/unset (length !== {@link PREFERENCE_MATRIX_LENGTH}):
 * morning 8–11AM → 1 (high), afternoon 2–5PM → 0.5 (medium), evening 7–10PM →
 * 0.2 (low), everything else → 0 (neutral, never negative).
 */
export function defaultPreferenceMatrix(): number[] {
  const matrix = new Array<number>(PREFERENCE_MATRIX_LENGTH).fill(0);
  for (let wd = 1; wd <= 7; wd++) {
    for (let hour = 8; hour < 11; hour++) matrix[matrixIndex(wd, hour)] = 1;
    for (let hour = 14; hour < 17; hour++) matrix[matrixIndex(wd, hour)] = 0.5;
    for (let hour = 19; hour < 22; hour++) matrix[matrixIndex(wd, hour)] = 0.2;
  }
  return matrix;
}

/** The stored matrix if well-formed, else the {@link defaultPreferenceMatrix} fallback. */
export function effectivePreferenceMatrix(prefMatrix: number[]): number[] {
  return prefMatrix.length === PREFERENCE_MATRIX_LENGTH
    ? prefMatrix
    : defaultPreferenceMatrix();
}

/** Preference-matrix value of the hour bucket `instant` falls in, in `timezone`. */
export function preferenceScoreAt(
  matrix: number[],
  instant: Date,
  timezone: string,
): number {
  const dateStr = localDateStr(instant, timezone);
  const wd = isoWeekday(dateStr);
  const hour = Math.floor(utcToMinutes(instant, timezone) / 60);
  return matrix[matrixIndex(wd, hour)] ?? 0;
}

/** Flat matrix index of the local hour bucket containing `atMs`. */
function cellIndexAt(atMs: number, timezone: string): number {
  return matrixIndex(
    isoWeekday(localDateStr(new Date(atMs), timezone)),
    Math.floor(utcToMinutes(new Date(atMs), timezone) / 60),
  );
}

/**
 * Nudges the ONE hour bucket containing `atMs` by `PREFERENCE_LEARNING_RATE ·
 * delta` (`delta` in `[-1, 1]`), clamped to `[-1, 1]`. Returns a NEW array.
 */
export function reinforcePreferenceCell(
  matrix: number[],
  atMs: number,
  timezone: string,
  delta: number,
  rate: number = PREFERENCE_LEARNING_RATE,
): number[] {
  const next = [...effectivePreferenceMatrix(matrix)];
  const idx = cellIndexAt(atMs, timezone);
  next[idx] = clamp(next[idx] + rate * delta, -1, 1);
  return next;
}

/**
 * First move of a placed session: lowers the old hour and raises the new hour
 * by `η·g`, `g = -dragDistanceReward(drag)` in `[0, 1]`. No-op for a zero-distance
 * move or one inside a single hour bucket. Returns a NEW array.
 */
export function reinforcePreferenceMove(
  matrix: number[],
  oldStartMs: number,
  newStartMs: number,
  timezone: string,
  dragDistanceMinutes: number,
  rate: number = PREFERENCE_LEARNING_RATE,
): number[] {
  const grade = -dragDistanceReward(dragDistanceMinutes);
  const sameCell =
    cellIndexAt(oldStartMs, timezone) === cellIndexAt(newStartMs, timezone);
  if (grade === 0 || sameCell) return [...effectivePreferenceMatrix(matrix)];

  const lowered = reinforcePreferenceCell(
    matrix,
    oldStartMs,
    timezone,
    -grade,
    rate,
  );
  return reinforcePreferenceCell(lowered, newStartMs, timezone, grade, rate);
}

File: services/bandit/src/core/preference.py
"""Preference matrix helpers + decay (port of ``preference.ts`` / ``matrix-decay.ts``).

The matrix is a flat 168-float array: 7 ISO weekdays x 24 hour buckets.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import ArrayLike, NDArray

from .constants import (
    MATRIX_HALF_LIFE_DAYS,
    PREFERENCE_LEARNING_RATE,
    PREFERENCE_MATRIX_LENGTH,
    PREFERENCE_SLOTS_PER_DAY,
)
from .reward import drag_distance_reward
from .slot import iso_weekday, local_date_str, utc_to_minutes


def matrix_index(iso_weekday_num: int, hour: int) -> int:
    return (iso_weekday_num - 1) * PREFERENCE_SLOTS_PER_DAY + hour


def default_preference_matrix() -> NDArray[np.float64]:
    m = np.zeros(PREFERENCE_MATRIX_LENGTH)
    for wd in range(1, 8):
        m[matrix_index(wd, 8) : matrix_index(wd, 11)] = 1.0
        m[matrix_index(wd, 14) : matrix_index(wd, 17)] = 0.5
        m[matrix_index(wd, 19) : matrix_index(wd, 22)] = 0.2
    return m


def effective_preference_matrix(pref: ArrayLike) -> NDArray[np.float64]:
    arr = np.asarray(pref, dtype=np.float64)
    if arr.shape == (PREFERENCE_MATRIX_LENGTH,):
        return arr
    return default_preference_matrix()


def cell_index_at(at_ms: int, timezone: str) -> int:
    return matrix_index(
        iso_weekday(local_date_str(at_ms, timezone)),
        utc_to_minutes(at_ms, timezone) // 60,
    )


def preference_score_at(matrix: ArrayLike, at_ms: int, timezone: str) -> float:
    return float(np.asarray(matrix)[cell_index_at(at_ms, timezone)])


def reinforce_preference_cell(
    matrix: ArrayLike,
    at_ms: int,
    timezone: str,
    delta: float,
    rate: float = PREFERENCE_LEARNING_RATE,
) -> NDArray[np.float64]:
    nxt = effective_preference_matrix(matrix).copy()
    idx = cell_index_at(at_ms, timezone)
    nxt[idx] = min(max(nxt[idx] + rate * delta, -1.0), 1.0)
    return nxt


def reinforce_preference_move(
    matrix: ArrayLike,
    old_start_ms: int,
    new_start_ms: int,
    timezone: str,
    drag_distance_minutes: float,
    rate: float = PREFERENCE_LEARNING_RATE,
) -> NDArray[np.float64]:
    grade = -drag_distance_reward(drag_distance_minutes)
    same = cell_index_at(old_start_ms, timezone) == cell_index_at(
        new_start_ms, timezone
    )
    if grade == 0 or same:
        return effective_preference_matrix(matrix).copy()
    lowered = reinforce_preference_cell(matrix, old_start_ms, timezone, -grade, rate)
    return reinforce_preference_cell(lowered, new_start_ms, timezone, grade, rate)


def decay_matrix(
    matrix: ArrayLike,
    delta_days: float,
    half_life_days: float = MATRIX_HALF_LIFE_DAYS,
) -> NDArray[np.float64]:
    arr = np.asarray(matrix, dtype=np.float64)
    if not delta_days > 0 or not half_life_days > 0:
        return arr.copy()
    return np.asarray(arr * 2.0 ** (-delta_days / half_life_days), dtype=np.float64)
